src/Build/BackEnd/Components/Logging/LoggingService.cs

```diff
@@ -546,9 +546,9 @@ public ICollection<string> GetWarningsToBeLoggedAsErrorsByProject(BuildEventCont
                 return null;
             }
 
-            if (_warningsAsMessagesByProject != null)
+            if (_warningsAsMessagesByProject != null && _warningsAsMessagesByProject.TryGetValue(key, out ISet<string> warningsAsMessages))
             {
-                warningsAsErrorsExcludingMessages.ExceptWith(_warningsAsMessagesByProject[key]);
+                warningsAsErrorsExcludingMessages.ExceptWith(warningsAsMessages);
 
                 // A non-null empty set means all warnings are errors. Avoid this.
                 if (warningsAsErrorsExcludingMessages.Count == 0)
```

Comment:
This probably belongs to outer scope after if from line 549, as relying on _warningsAsErrorsByProject returning non empty set is a fragile promise.